### PR TITLE
Mandatory Elements To Rerender For Strategies

### LINKED ISSUE #6372: Possibly add a requirement for `elementsToRerender` in canvas strategy output
When adding new canvas strategies it is very easy to forget to use the `setElementsToRerenderCommand` when required (which is most of the time). The `CustomStrategyState` already has a field `elementsToRerender`. It might make more sense to remove the command, and to bring `elementsToRerender` up into the `StrategyApplicationResult` to enforce a conscious decision around whether elements need to be re-rendered as a result of applying this strategy.

### PR DESCRIPTION
**Pull request by seanparsons**
**Problem:**
When adding new canvas strategies it is very easy to forget to use the `setElementsToRerenderCommand` when required (which is most of the time).

**Fix:**
Now instead of commands being the source of the elements to rerender, now that value is returned specifically by the strategy and is required to be passed to `strategyApplicationResult`. That value is then applied to the editor state taking it from the `StrategyApplicationResult`.

**Commit Details:**
- Added `combineElementsToRerender` utility function.
- Added `appendElementsToRerenderToApplyResult` utility function.
- Added `StrategyApplicationResult.elementsToRerender`.
- Removed `setElementsToRerenderCommand` and `appendElementsToRerenderCommand` and the associated supporting code.
- Moved values originally passed to `setElementsToRerenderCommand` to the new parameter that `strategyApplicationResult` now has.
- Moved values originally passed to `appendElementsToRerenderCommand` to a call to `appendElementsToRerenderToApplyResult`.
- Added `applyElementsToRerenderFromStrategyResult` utility function.
- Hooked in the `applyElementsToRerenderFromStrategyResult` function
  into various strategy functions to update the `EditorState`.

**Manual Tests:**
I hereby swear that:

- [x] I opened a hydrogen project and it loaded
- [x] I could navigate to various routes in Preview mode

Fixes #6372

### CODE REVIEW

File: editor/src/components/canvas/canvas-strategies/strategies/absolute-resize-bounding-box-strategy.tsx [hunk 1]
```diff
@@ -289,19 +289,24 @@ export function absoluteResizeBoundingBoxStrategy(
...
-              setElementsToRerenderCommand(retargetedTargets),
-            ])
+            return strategyApplicationResult(
+              [
+                ...commandsForSelectedElements,
+                setSnappingGuidelines('mid-interaction', guidelinesWithSnappingVector),
+                updateHighlightedViews('mid-interaction', []),
+                setCursorCommand(pickCursorFromEdgePosition(edgePosition)),
+              ],
+              'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/absolute-resize-bounding-box-strategy.tsx:299]: this should be retargetedTargets based on the diff

File: editor/src/components/canvas/canvas-strategies/strategies/convert-to-absolute-and-move-strategy.tsx [hunk 2]
```diff
@@ -278,10 +278,10 @@ function convertToAbsoluteAndMoveStrategyFactory(setHuggingParentToFixed: SetHug
             },
           })
 
-          return strategyApplicationResult([
-            ...absoluteMoveApplyResult.commands,
-            strategyIndicatorCommand,
-          ])
+          return strategyApplicationResult(
+            [...absoluteMoveApplyResult.commands, strategyIndicatorCommand],
+            'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/convert-to-absolute-and-move-strategy.tsx:283]: looks like this strategy forgot to set elementsToRerender, we should fix it!

File: editor/src/components/canvas/canvas-strategies/strategies/drag-to-insert-metastrategy.tsx [hunk 3]
```diff
@@ -280,6 +281,7 @@ function dragToInsertStrategyFactory(
               reparentCommand,
               ...optionalWrappingCommand,
             ],
+            'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/drag-to-insert-metastrategy.tsx:284]: looks like another place to fix

File: editor/src/components/canvas/canvas-strategies/strategies/drag-to-move-metastrategy.tsx [hunk 4]
```diff
@@ -153,23 +153,26 @@ export function doNothingStrategy(canvasState: InteractionCanvasState): CanvasSt
...
               },
             },
-          },
-        }),
-        setCursorCommand(CSSCursor.NotPermitted),
-      ])
+          }),
+          setCursorCommand(CSSCursor.NotPermitted),
+        ],
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/drag-to-move-metastrategy.tsx:174]: or maybe []?

File: editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx [hunk 5]
```diff
@@ -348,6 +349,7 @@ export function drawToInsertStrategyFactory(
 
               return strategyApplicationResult(
                 [insertionCommand.command, reparentCommand, ...optionalWrappingCommand],
+                'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx:352]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx [hunk 6]
```diff
@@ -303,6 +303,7 @@ export function drawToInsertStrategyFactory(
                   resizeCommand,
                   ...optionalWrappingCommand,
                 ],
+                'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx:306]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-text-strategy.tsx [hunk 7]
```diff
@@ -107,24 +107,27 @@ export const drawToInsertTextMetaStrategy: MetaCanvasStrategy = (
...
-                  'existing',
-                  'no-text-selection',
-                ),
-              },
-            }),
-          ])
+                },
+              }),
+            ],
+            'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-text-strategy.tsx:129]: definitely looks like one to fix

File: editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx [hunk 8]
```diff
@@ -165,6 +165,7 @@ const gridDrawToInsertStrategyInner =
               showGridControls('mid-interaction', targetParent),
               updateHighlightedViews('mid-interaction', [targetParent]),
             ],
+            'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx:168]: or []?

File: editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx [hunk 9]
```diff
@@ -258,6 +259,7 @@ const gridDrawToInsertStrategyInner =
                 : null,
             ]),
           ],
+          'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx:262]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/grid-rearrange-keyboard-strategy.ts [hunk 10]
```diff
@@ -133,6 +133,7 @@ export function gridRearrangeResizeKeyboardStrategy(
           gridRowStart,
           gridRowEnd,
         }),
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/grid-rearrange-keyboard-strategy.ts:136]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/grid-rearrange-move-strategy.ts [hunk 11]
```diff
@@ -155,6 +155,7 @@ export const gridRearrangeMoveStrategy: CanvasStrategyFactory = (
 
       return strategyApplicationResult(
         [...midInteractionCommands, ...onCompleteCommands, ...commands],
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/grid-rearrange-move-strategy.ts:158]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/grid-resize-element-strategy.ts [hunk 12]
```diff
@@ -140,6 +140,7 @@ export const gridResizeElementStrategy: CanvasStrategyFactory = (
 
       return strategyApplicationResult(
         setGridPropsCommands(selectedElement, gridTemplate, gridPropsWithDragOver(gridProps)),
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/grid-resize-element-strategy.ts:143]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/keyboard-set-font-weight-strategy.tsx [hunk 13]
```diff
@@ -87,7 +87,7 @@ export function keyboardSetFontWeightStrategy(
         ),
       )
 
-      return strategyApplicationResult(commands)
+      return strategyApplicationResult(commands, 'rerender-all-elements')
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/keyboard-set-font-weight-strategy.tsx:90]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/keyboard-set-opacity-strategy.tsx [hunk 14]
```diff
@@ -57,7 +57,7 @@ export function keyboardSetOpacityStrategy(
         setProperty('always', path, PP.create('style', 'opacity'), inputValue),
       )
 
-      return strategyApplicationResult(commands)
+      return strategyApplicationResult(commands, 'rerender-all-elements')
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/keyboard-set-opacity-strategy.tsx:60]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/rearrange-grid-swap-strategy.ts [hunk 15]
```diff
@@ -121,7 +121,7 @@ export const rearrangeGridSwapStrategy: CanvasStrategyFactory = (
         return emptyStrategyApplicationResult
       }
 
-      return strategyApplicationResult(commands)
+      return strategyApplicationResult(commands, 'rerender-all-elements')
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/rearrange-grid-swap-strategy.ts:124]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx [hunk 16]
```diff
@@ -348,6 +347,7 @@ export function applyStaticReparent(
 
             return {
               commands: [...midInteractionCommands, ...interactionFinishCommands],
+              elementsToRerender: 'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx:350]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/resize-grid-strategy.ts [hunk 17]
```diff
@@ -189,10 +191,9 @@ export const resizeGridStrategy: CanvasStrategyFactory = (
           ),
           propertyValueAsString,
         ),
-        setElementsToRerenderCommand([gridPath]),
       ]
 
-      return strategyApplicationResult(commands)
+      return strategyApplicationResult(commands, 'rerender-all-elements')
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/resize-grid-strategy.ts:196]: this looks like this should be `[gridPath]`

File: editor/src/components/canvas/canvas-strategies/strategies/set-flex-gap-strategy.tsx [hunk 18]
```diff
@@ -172,30 +172,33 @@ export const setFlexGapStrategy: CanvasStrategyFactory = (
       }
 
       if (shouldTearOffGap) {
-        return strategyApplicationResult([
-          deleteProperties('always', selectedElement, [StyleGapProp]),
-        ])
+        return strategyApplicationResult(
+          [deleteProperties('always', selectedElement, [StyleGapProp])],
+          'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/set-flex-gap-strategy.tsx:177]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/set-grid-gap-strategy.tsx [hunk 19]
```diff
@@ -171,30 +171,33 @@ export const setGridGapStrategy: CanvasStrategyFactory = (
         axis === 'row' ? updatedGridGapMeasurement.row : updatedGridGapMeasurement.column
 
       if (shouldTearOffGapByAxis) {
-        return strategyApplicationResult([
-          deleteProperties('always', selectedElement, [axisStyleProp]),
-        ])
+        return strategyApplicationResult(
+          [deleteProperties('always', selectedElement, [axisStyleProp])],
+          'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/set-grid-gap-strategy.tsx:176]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx [hunk 20]
```diff
@@ -246,11 +245,70 @@ export const setPaddingStrategy: CanvasStrategyFactory = (canvasState, interacti
...
+              selectedElements.map((path) => ({
+                action: 'set-padding',
+                target: activeFrameTargetPath(path),
+                source: zeroRectIfNullOrInfinity(
+                  MetadataUtils.getFrameInCanvasCoords(path, canvasState.startingMetadata),
+                ),
+              })),
+            ),
+          ],
+          'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx:273]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx [hunk 21]
```diff
@@ -246,11 +245,70 @@ export const setPaddingStrategy: CanvasStrategyFactory = (canvasState, interacti
...
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx:301]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx [hunk 22]
```diff
@@ -269,59 +327,9 @@ export const setPaddingStrategy: CanvasStrategyFactory = (canvasState, interacti
...
-            action: 'set-padding',
-            target: activeFrameTargetPath(path),
-            source: zeroRectIfNullOrInfinity(
-              MetadataUtils.getFrameInCanvasCoords(path, canvasState.startingMetadata),
-            ),
-          })),
-        ),
-      ])
+        ],
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx:331]: possibly one to fix here too

File: editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts [hunk 23]
```diff
@@ -152,17 +152,20 @@ export function applyMoveCommon(
...
+          pushIntendedBoundsAndUpdateGroups(
+            commandsForSelectedElements.intendedBounds,
+            'starting-metadata',
+          ),
+          updateHighlightedViews('mid-interaction', []),
+
+          setCursorCommand(CSSCursor.Select),
+          setActiveFrames(getActiveFrames(commandsForSelectedElements.intendedBounds)),
+        ],
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts:167]: based on the diff I think this should be `targets`

File: editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts [hunk 24]
```diff
@@ -189,18 +192,21 @@ export function applyMoveCommon(
...
+          setSnappingGuidelines('mid-interaction', guidelinesWithSnappingVector),
+          pushIntendedBoundsAndUpdateGroups(
+            commandsForSelectedElements.intendedBounds,
+            'starting-metadata',
+          ),
+
+          setCursorCommand(CSSCursor.Select),
+          setActiveFrames(getActiveFrames(commandsForSelectedElements.intendedBounds)),
+        ],
+        'rerender-all-elements',
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts:208]: based on the diff it looks like this should be [...targets, ...targetsForSnapping]

File: editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts [hunk 25]
```diff
@@ -354,7 +360,7 @@ export function moveInspectorStrategy(
         intendedBounds.push(...moveCommandsResult.intendedBounds)
       }
       commands.push(pushIntendedBoundsAndUpdateGroups(intendedBounds, 'live-metadata'))
-      commands.push(setElementsToRerenderCommand(selectedElementPaths))
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts:357]: do these removed setElementsToRerenderCommands have a counterpart which is setting the elementsToRerender value?

Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts:357]: They already appear to from some surrounding logic which also passes the selected paths through as the elements to render.

File: editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts (hunk 24 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/shared-move-strategies-helpers.ts]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/resize-grid-strategy.ts (hunk 17 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/resize-grid-strategy.ts]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx (hunk 16 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx (hunk 8 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/drag-to-move-metastrategy.tsx (hunk 4 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/drag-to-move-metastrategy.tsx]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/ancestor-metastrategy.tsx [hunk 26]
```diff
@@ -125,11 +129,9 @@ export function ancestorMetaStrategy(
       return nextAncestorResult.map((s) => ({
         ...s,
         fitness: fitness(s),
-        apply: appendCommandsToApplyResult(
-          s.apply,
-          [appendElementsToRerenderCommand([target])],
-          [],
-        ),
+        apply: appendElementsToRerenderToApplyResult(appendCommandsToApplyResult(s.apply, [], []), [
```
[the remark this answers is not in the document]
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/ancestor-metastrategy.tsx]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/absolute-resize-bounding-box-strategy.tsx (hunk 1 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/absolute-resize-bounding-box-strategy.tsx]: Fixed in 71214dfb2.

File: editor/src/components/canvas/canvas-strategies/strategies/convert-to-absolute-and-move-strategy.tsx (hunk 2 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/convert-to-absolute-and-move-strategy.tsx:286]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/drag-to-insert-metastrategy.tsx (hunk 3 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/drag-to-insert-metastrategy.tsx:287]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx (hunk 5 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx:358]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx (hunk 6 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-metastrategy.tsx:309]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-text-strategy.tsx (hunk 7 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/draw-to-insert-text-strategy.tsx:132]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx (hunk 9 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/grid-draw-to-insert-strategy.tsx:265]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/grid-rearrange-keyboard-strategy.ts (hunk 10 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/grid-rearrange-keyboard-strategy.ts:139]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/grid-resize-element-strategy.ts (hunk 12 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/grid-resize-element-strategy.ts:146]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/keyboard-set-opacity-strategy.tsx (hunk 14 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/keyboard-set-opacity-strategy.tsx]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/rearrange-grid-swap-strategy.ts (hunk 15 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/rearrange-grid-swap-strategy.ts]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/set-flex-gap-strategy.tsx (hunk 18 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/set-flex-gap-strategy.tsx:180]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/set-grid-gap-strategy.tsx (hunk 19 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/set-grid-gap-strategy.tsx:179]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx (hunk 20 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx:276]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx (hunk 21 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx:307]: Added comment about follow up in 38ec355d5.

File: editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx (hunk 22 above)
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/strategies/set-padding-strategy.tsx:340]: Added comment about follow up in 38ec355d5.

**Pull request merged**